File: js/packages/teams-ai/src/OpenAIPlanner.ts
/* eslint-disable security/detect-object-injection */
/**
 * @module teams-ai
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import { PredictedDoCommand, Planner, Plan } from './Planner';
import { TurnState } from './TurnState';
import { DefaultTempState, DefaultTurnState } from './DefaultTurnStateManager';
import { TurnContext } from 'botbuilder';
import {
    OpenAIClient,
    OpenAIClientResponse,
    CreateCompletionRequest,
    CreateCompletionResponse,
    CreateChatCompletionRequest,
    CreateChatCompletionResponse,
    ChatCompletionRequestMessage
} from './OpenAIClients';
import { ResponseParser } from './ResponseParser';
import { ConversationHistory } from './ConversationHistory';
import { AI, ConfiguredAIOptions } from './AI';
import { PromptTemplate } from './Prompts';

/**
 * Options for the OpenAI based planner.
 */
export interface OpenAIPlannerOptions {
    /**
     * OpenAI API key.
     */
    apiKey: string;

    /**
     * Optional. OpenAI organization.
     */
    organization?: string;

    /**
     * Optional. OpenAI endpoint.
     */
    endpoint?: string;

    /**
     * The default model to use.
     *
     * @remarks
     * Prompts can override this model.
     */
    defaultModel: string;

    /**
     * Optional. A flag indicating if the planner should only say one thing per turn.
     *
     * @remarks
     * The planner will attempt to combine multiple SAY commands into a single SAY command when true.
     * Defaults to false.
     */
    oneSayPerTurn?: boolean;

    /**
     * Optional. A flag indicating if the planner should use the 'system' role when calling OpenAI's
     * chatCompletion API.
     *
     * @remarks
     * The planner current uses the 'user' role by default as this tends to generate more reliable
     * instruction following. Defaults to false.
     */
    useSystemMessage?: boolean;

    /**
     * Optional. A flag indicating if the planner should log requests to the console.
     *
     * @remarks
     * Both the prompt text and the completion response will be logged to the console. For
     * chatCompletion calls the outgoing messages array will also be logged.
     * Defaults to false.
     */
    logRequests?: boolean;
}

/**
 * A planner that uses OpenAI's textCompletion and chatCompletion API's to generate plans.
 *
 * @remarks
 * This planner can be configured to use different models for different prompts. The prompts model
 * will determine which API is used to generate the plan. Any model that starts with 'gpt-' will
 * use the chatCompletion API, otherwise the textCompletion API will be used.
 * @template TState Optional. Type of the applications turn state.
 * @template TOptions Optional. Type of the planner options.
 */
export class OpenAIPlanner<
    TState extends TurnState = DefaultTurnState,
    TOptions extends OpenAIPlannerOptions = OpenAIPlannerOptions
> implements Planner<TState>
{
    private readonly _options: TOptions;
    private readonly _client: OpenAIClient;

    /**
     * Creates a new instance of the OpenAI based planner.
     *
     * @param options Options for the OpenAI based planner.
     */
    public constructor(options: TOptions) {
        this._options = Object.assign(
            {
                oneSayPerTurn: false,
                useSystemMessage: false,
                logRequests: false
            } as TOptions,
            options
        );
        this._client = this.createClient(this._options);
    }

    /**
     * Returns the configured options for the planner.
     */
    public get options(): TOptions {
        return this._options;
    }

    /**
     * Completes a prompt without returning a plan.
     *
     * @param context Context for the current turn of conversation.
     * @param state Application state for the current turn of conversation.
     * @param prompt Prompt to complete.
     * @param options Configuration options for the AI system.
     * @returns The response from the prompt. Can return undefined to indicate the prompt was rate limited.
     */
    public async completePrompt(
        context: TurnContext,
        state: TState,
        prompt: PromptTemplate,
        options: ConfiguredAIOptions<TState>
    ): Promise<string> {
        // Check for chat completion model
        const model = this.getModel(prompt);
        if (model.startsWith('gpt-')) {
            // Request base chat completion
            const temp = (state['temp']?.value ?? {}) as DefaultTempState;
            const chatRequest = this.createChatCompletionRequest(state, prompt, temp.input, options);
            const result = await this.createChatCompletion(chatRequest);
            return result.data?.choices[0]?.message?.content ?? '';
        } else {
            // Request base prompt completion
            const promptRequest = this.createCompletionRequest(prompt);
            const result = await this.createCompletion(promptRequest);
            return result.data?.choices[0]?.text ?? '';
        }
    }

    /**
     * Completes a prompt and generates a plan for the AI system to execute.
     *
     * @param context Context for the current turn of conversation.
     * @param state Application state for the current turn of conversation.
     * @param prompt Prompt to complete.
     * @param options Configuration options for the AI system.
     * @returns The plan that was generated.
     */
    public async generatePlan(
        context: TurnContext,
        state: TState,
        prompt: PromptTemplate,
        options: ConfiguredAIOptions<TState>
    ): Promise<Plan> {
        // Check for chat completion model
        let status: number;
        let response: string;
        const model = this.getModel(prompt);
        if (model.startsWith('gpt-')) {
            // Request base chat completion
            const temp = (state['temp']?.value ?? {}) as DefaultTempState;
            const chatRequest = await this.createChatCompletionRequest(state, prompt, temp.input, options);
            const result = await this.createChatCompletion(chatRequest);
            status = result?.status;
            response = status === 200 ? result.data?.choices[0]?.message?.content ?? '' : '';
        } else {
            // Request base prompt completion
            const promptRequest = this.createCompletionRequest(prompt);
            const result = await this.createCompletion(promptRequest);
            status = result?.status;
            response = status === 200 ? result.data?.choices[0]?.text ?? '' : '';
        }

        // Ensure we weren't rate limited
        if (status === 429) {
            return {
                type: 'plan',
                commands: [
                    {
                        type: 'DO',
                        action: AI.RateLimitedActionName,
                        entities: {}
                    } as PredictedDoCommand
                ]
            };
        }

        // Parse returned prompt response
        if (response) {
            // Patch the occasional "Then DO" which gets predicted
            response = response.trim().replace('Then DO ', 'THEN DO ').replace('Then SAY ', 'THEN SAY ');
            if (response.startsWith('THEN ')) {
                response = response.substring(5);
            }

            // Remove response prefix
            if (options.history.assistantPrefix) {
                // The model sometimes predicts additional text for the human side of things so skip that.
                const pos = response.toLowerCase().indexOf(options.history.assistantPrefix.toLowerCase());
                if (pos >= 0) {
                    response = response.substring(pos + options.history.assistantPrefix.length);
                }
            }

            // Parse response into commands
            const plan = ResponseParser.parseResponse(response.trim());

            // Filter to only a single SAY command
            if (this._options.oneSayPerTurn) {
                let spoken = false;
                plan.commands = plan.commands.filter((cmd) => {
                    if (cmd.type == 'SAY') {
                        if (spoken) {
                            return false;
                        }

                        spoken = true;
                    }

                    return true;
                });
            }

            return plan;
        }

        // Return an empty plan by default
        return { type: 'plan', commands: [] };
    }

    /**
     * @param options
     * @private
     */
    protected createClient(options: TOptions): OpenAIClient {
        return new OpenAIClient({
            apiKey: options.apiKey,
            organization: options.organization,
            endpoint: options.endpoint
        });
    }

    /**
     * @param prompt
     * @private
     */
    private getModel(prompt: PromptTemplate): string {
        if (Array.isArray(prompt.config.default_backends) && prompt.config.default_backends.length > 0) {
            return prompt.config.default_backends[0];
        } else {
            return this._options.defaultModel;
        }
    }

    /**
     * @param state
     * @param prompt
     * @param userMessage
     * @param options
     * @private
     */
    private createChatCompletionRequest(
        state: TState,
        prompt: PromptTemplate,
        userMessage: string,
        options: ConfiguredAIOptions<TState>
    ): CreateChatCompletionRequest {
        // Clone prompt config
        const request: CreateChatCompletionRequest = Object.assign(
            {
                model: this.getModel(prompt),
                messages: []
            },
            prompt.config.completion as CreateChatCompletionRequest
        );
        this.patchStopSequences(request);

        // Populate system message
        request.messages.push({
            role: this._options.useSystemMessage ? 'system' : 'user',
            content: prompt.text
        });

        // Populate conversation history
        if (options.history.trackHistory) {
            const userPrefix = options.history.userPrefix.trim().toLowerCase();
            const assistantPrefix = options.history.assistantPrefix.trim().toLowerCase();
            const history = ConversationHistory.toArray(state, options.history.maxTokens);
            for (let i = 0; i < history.length; i++) {
                let line = history[i];
                const lcLine = line.toLowerCase();
                if (lcLine.startsWith(userPrefix)) {
                    line = line.substring(userPrefix.length).trim();
                    request.messages.push({
                        role: 'user',
                        content: line
                    });
                } else if (lcLine.startsWith(assistantPrefix)) {
                    line = line.substring(assistantPrefix.length).trim();
                    request.messages.push({
                        role: 'assistant',
                        content: line
                    });
                }
            }
        }

        // Add user message
        if (userMessage) {
            request.messages.push({
                role: 'user',
                content: userMessage
            });
        }

        return request;
    }

    /**
     * @param prompt
     * @private
     */
    private createCompletionRequest(prompt: PromptTemplate): CreateCompletionRequest {
        // Clone prompt config
        const request: CreateCompletionRequest = Object.assign({}, prompt.config.completion as CreateCompletionRequest);
        this.patchStopSequences(request);
        request.model = this.getModel(prompt);
        request.prompt = prompt.text;
        return request;
    }

    /**
     * @param request
     * @private
     */
    private patchStopSequences(request: any): void {
        if (request['stop_sequences']) {
            request.stop = request['stop_sequences'];
            delete request['stop_sequences'];
        }
    }

    /**
     * @param request
     * @private
     */
    private async createChatCompletion(
        request: CreateChatCompletionRequest
    ): Promise<OpenAIClientResponse<CreateChatCompletionResponse>> {
        let response: OpenAIClientResponse<CreateChatCompletionResponse>;
        let error: { status?: number } = {};
        const startTime = new Date().getTime();
        try {
            response = await this._client.createChatCompletion(request);
        } catch (err: any) {
            error = err;
            throw err;
        } finally {
            if (this._options.logRequests) {
                const duration = new Date().getTime() - startTime;
                console.log(`\nCHAT REQUEST:\n\`\`\`\n${printChatMessages(request.messages)}\`\`\``);
                if (response!) {
                    if (response.status != 429) {
                        const choice =
                            Array.isArray(response?.data?.choices) &&
                            response.data &&
                            response.data.choices.length > 0 &&
                            response.data.choices[0].message?.content;
                        // TODO: if we add telemetry, we should log this
                        console.log(
                            `CHAT SUCCEEDED: status=${response.status} duration=${duration} prompt=${response?.data?.usage?.prompt_tokens} completion=${response?.data?.usage?.completion_tokens} response=${choice}`
                        );
                    } else {
                        console.error(
                            `CHAT FAILED due to rate limiting: status=${
                                response.status
                            } duration=${duration} headers=${JSON.stringify(response.headers)}`
                        );
                    }
                } else {
                    console.error(
                        `CHAT FAILED: status=${error?.status} duration=${duration} message=${error?.toString()}`
                    );
                }
            }
        }

        return response!;
    }

    /**
     * @param request
     * @private
     */
    private async createCompletion(
        request: CreateCompletionRequest
    ): Promise<OpenAIClientResponse<CreateCompletionResponse>> {
        let response: OpenAIClientResponse<CreateCompletionResponse>;
        let error: { status?: number } = {};
        const startTime = new Date().getTime();
        try {
            response = await this._client.createCompletion(request);
        } catch (err: any) {
            error = err;
            throw err;
        } finally {
            if (this._options.logRequests) {
                const duration = new Date().getTime() - startTime;
                console.log(`\nPROMPT REQUEST:\n\`\`\`\n${request.prompt}\`\`\``);
                if (response!) {
                    if (response.status != 429) {
                        const choice =
                            Array.isArray(response?.data?.choices) && response.data && response.data.choices.length > 0
                                ? response.data.choices[0].text
                                : '';
                        // TODO: telemetry
                        console.log(
                            `PROMPT SUCCEEDED: status=${response.status} duration=${duration} prompt=${
                                response.data!.usage?.prompt_tokens
                            } completion=${response.data!.usage?.completion_tokens} response=${choice}`
                        );
                    } else {
                        console.error(
                            `PROMPT FAILED: status=${response.status} duration=${duration} headers=${JSON.stringify(
                                response.headers
                            )}`
                        );
                    }
                } else {
                    console.error(
                        `PROMPT FAILED: status=${error?.status} duration=${duration} message=${error?.toString()}`
                    );
                }
            }
        }

        return response!;
    }
}

/**
 * @param messages
 * @private
 */
function printChatMessages(messages: ChatCompletionRequestMessage[]): string {
    let text = '';
    messages.forEach((msg) => {
        switch (msg.role) {
            case 'system':
                text += msg.content + '\n';
                break;
            default:
                text += `\n${msg.role}: ${msg.content}`;
                break;
        }
    });

    return text;
}

File: python/packages/ai/teams/ai/planner/openai_planner.py
"""
Copyright (c) Microsoft Corporation. All rights reserved.
Licensed under the MIT License.
"""
from datetime import datetime
from typing import Any, Callable, List, Tuple, Union

import semantic_kernel as sk
from botbuilder.core import TurnContext
from semantic_kernel.connectors.ai import (
    ChatCompletionClientBase,
    ChatRequestSettings,
    CompleteRequestSettings,
    TextCompletionClientBase,
)
from semantic_kernel.connectors.ai.open_ai import (
    OpenAIChatCompletion,
    OpenAITextCompletion,
)

from teams.ai.ai_history_options import AIHistoryOptions
from teams.ai.prompts import PromptManager
from teams.ai.prompts.prompt_template import PromptTemplate
from teams.ai.prompts.utils import generate_sk_prompt_template_config
from teams.ai.state import ConversationT, TempT, TurnState, UserT

from .command_type import CommandType
from .conversation_history import state_to_history_array
from .openai_planner_options import OpenAIPlannerOptions
from .plan import Plan
from .planner import Planner
from .response_parser import parse_response

# Define role names used by OpenAI API
SYSTEM_ROLE_NAME: str = "system"
USER_ROLE_NAME: str = "user"
ASSISTANT_ROLE_NAME: str = "assistant"


class OpenAIPlanner(Planner):
    """
    Planner that uses OpenAI's API to generate prompts
    """

    _options: OpenAIPlannerOptions
    _sk: sk.Kernel
    _prompt_manager: PromptManager

    def __init__(self, options: OpenAIPlannerOptions) -> None:
        self._options = options
        self._prompt_manager = PromptManager(options.prompt_folder)
        self._sk = sk.Kernel()
        self._add_text_completion_service()
        self._add_chat_completion_service()

    async def add_function(
        self, name: str, handler: Callable[[TurnContext, TurnState], Any], *, allow_overrides=False
    ) -> None:
        self._prompt_manager.add_function(name, handler, allow_overrides=allow_overrides)

    async def generate_plan(
        self,
        turn_context: TurnContext,
        state: TurnState,
        prompt_name_or_template: Union[str, PromptTemplate],
        *,
        history_options: AIHistoryOptions = AIHistoryOptions(),
    ) -> Plan:
        """
        Generates a plan based on the given turn state and prompt name or template.
        Args:
            turn_context (TurnContext): The turn context for current turn of conversation
            state (TurnState): The current turn state.
            prompt_name_or_template (Union[str, PromptTemplate]):
                The name of the prompt or a prompt template to use.
            history_options (AIHistoryOptions): The options for the AI history.
        Returns:
            Plan: The generated plan.
        """
        prompt_template = await self._prompt_manager.render_prompt(
            turn_context, state, prompt_name_or_template
        )
        result = await self._complete_prompt(state, prompt_template, history_options)

        if len(result) > 0:
            # Patch the occasional "Then DO" which gets predicted
            result = result.strip().replace("Then DO ", "THEN DO ").replace("Then SAY ", "THEN SAY")
            if result.startswith("THEN "):
                result = result[5:]

            assistant_prefix = (
                history_options.assistant_prefix if history_options is not None else None
            )

            if assistant_prefix:
                # The model sometimes predicts additional text
                # for the human side of things so skip that.
                position = result.lower().find(assistant_prefix.lower())
                if position >= 0:
                    result = result[position + len(assistant_prefix) :]

            plan: Plan = parse_response(result)

            # Filter to only a single SAY command
            if self._options.one_say_per_turn:
                spoken: bool = False
                new_commands = []
                for command in plan.commands:
                    if command.type == CommandType.SAY:
                        if spoken:
                            continue

                        new_commands.append(command)
                        spoken = True
                    else:
                        new_commands.append(command)
                plan.commands = new_commands

            return plan

        return Plan()

    def _add_text_completion_service(self) -> None:
        # TODO: default_model may not be text completion
        self._sk.add_text_completion_service(
            "openai_text_completion",
            OpenAITextCompletion(
                self._options.default_model, self._options.api_key, self._options.organization
            ),
        )

    def _add_chat_completion_service(self) -> None:
        # TODO: default_model may not be chat completion
        self._sk.add_chat_service(
            "openai_chat_completion",
            OpenAIChatCompletion(
                self._options.default_model, self._options.api_key, self._options.organization
            ),
        )

    async def _complete_prompt(
        self,
        turn_state: TurnState[ConversationT, UserT, TempT],
        prompt_template: PromptTemplate,
        history_options: AIHistoryOptions,
    ) -> str:
        model: str = self._get_model(prompt_template)
        is_chat_completion: bool = model.lower().startswith("gpt-")
        start_time = datetime.now()
        log_prefix = "CHAT" if is_chat_completion else "PROMPT"

        self._log_request(f"\n{log_prefix} REQUEST: \n'''\n{prompt_template.text}\n'''")

        if is_chat_completion:
            user_message = turn_state.temp.value.input
            result = await self._complete_chat(
                prompt_template, turn_state, user_message, history_options
            )
        else:
            result = await self._complete_text(prompt_template)

        duration = datetime.now() - start_time
        # TODO: investigate how to get prompt/completion tokens
        self._log_request(f"\n{log_prefix} SUCCEEDED: duration={duration} response={result}")

        return result

    async def _complete_chat(
        self,
        prompt_template: PromptTemplate,
        state: TurnState,
        user_message: str,
        options: AIHistoryOptions,
    ):
        prompt_template_config = generate_sk_prompt_template_config(prompt_template)
        request_settings = ChatRequestSettings.from_completion_config(
            prompt_template_config.completion
        )

        chat_completion_client = self._sk.get_ai_service(ChatCompletionClientBase)(self._sk)

        chat_history: List[Tuple[str, str]] = []
        if self._options.use_system_message:
            chat_history.append((SYSTEM_ROLE_NAME, prompt_template.text))
        else:
            chat_history.append((USER_ROLE_NAME, prompt_template.text))

        # Populate conversation history
        if options.track_history:
            user_prefix = options.user_prefix
            assistant_prefix = options.assistant_prefix
            history: List[str] = state_to_history_array(state, options.max_tokens)

            for line in history:
                if line.lower().startswith(user_prefix.lower()):
                    chat_history.append((USER_ROLE_NAME, line[len(user_prefix) :].strip()))
                elif line.lower().startswith(assistant_prefix.lower()):
                    chat_history.append(
                        (ASSISTANT_ROLE_NAME, line[len(assistant_prefix) :].strip())
                    )

        if user_message:
            chat_history.append((USER_ROLE_NAME, user_message))

        result = await chat_completion_client.complete_chat_async(chat_history, request_settings)
        if isinstance(result, str):
            return result

        return "\n".join(result)

    async def _complete_text(self, prompt_template: PromptTemplate):
        prompt_template_config = generate_sk_prompt_template_config(prompt_template)
        request_settings = CompleteRequestSettings.from_completion_config(
            prompt_template_config.completion
        )

        text_completion_client = self._sk.get_ai_service(TextCompletionClientBase)(self._sk)
        result = await text_completion_client.complete_async(prompt_template.text, request_settings)
        if isinstance(result, str):
            return result

        return "\n".join(result)

    def _get_model(self, prompt_template: PromptTemplate) -> str:
        default_backends = prompt_template.config.default_backends
        if default_backends and len(default_backends) > 0:
            return default_backends[0]

        return self._options.default_model

    def _log_request(self, request: str) -> None:
        if self._options.log_requests:
            print(request)
